File: source/downloads/code/LeetCode/Java/Binary-Tree-Zigzag-Level-Order-Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ret = LinkedList<List<Integer>>();
        travel(root,ret,0); 
    }

    public void travel(TreeNode node, List<List<Integer>> ret, int level){
        if(node==null) return ;

        if(ret.size() <= level){
            List<Integer> newLevel = new LinkedList<Integer>();
            ret.add(newLevel);
        }

        List<Integer> collection = ret.get(level);
        if(level%2 == 0)
            collection.add(node.val);
        else
            collection.add(0,node.val);

        travel(node.left, ret, level+1);
        travel(node.right, ret, level+1);

    }
}

File: downloads/code/LeetCode/Python/Binary-Tree-Zigzag-Level-Order-Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        ret = []
        stack = []
        level = 0
        stack.append(root)
        while stack:
            newstack = []
            if level % 2 == 0:
                ret.append([x.val for x in stack])
            else:
                ret.append([stack[x].val for x in range(len(stack)-1,-1,-1)])
            
            for node in stack:
                if node.left:
                    newstack.append(node.left)
                if node.right:
                    newstack.append(node.right)
            level += 1
            stack = newstack
        return ret
